cmd/openfga/openfga_test.go

```diff
@@ -0,0 +1,968 @@
+		},
+	}
+
+	ulid, err := id.NewString()
+	require.NoError(t, err)
+
+	name := fmt.Sprintf("openfga-%s", ulid)
+
+	ctx := context.Background()
+
+	cont, err := dockerClient.ContainerCreate(ctx, &containerCfg, &hostCfg, nil, nil, name)
+	require.NoError(t, err, "failed to create openfga docker container")
+
+	stopContainer := func() {
+
+		timeout := 5 * time.Second
+
+		err := dockerClient.ContainerStop(ctx, cont.ID, &timeout)
+		if err != nil && !client.IsErrNotFound(err) {
+			t.Fatalf("failed to stop openfga container: %v", err)
+		}
+	}
+
+	err = dockerClient.ContainerStart(ctx, cont.ID, types.ContainerStartOptions{})
+	require.NoError(t, err)
+
+	t.Cleanup(stopContainer)
+
+	// spin up a goroutine to survive any test panics or terminations to expire/stop the running container
+	go func() {
+		time.Sleep(2 * time.Minute)
+
+		// swallow the error because by this point we've terminated
+		_ = dockerClient.ContainerStop(ctx, cont.ID, nil)
+	}()
+
+	containerJSON, err := dockerClient.ContainerInspect(ctx, cont.ID)
+	require.NoError(t, err)
+
+	ports := containerJSON.NetworkSettings.Ports
+
+	m, ok := ports["8080/tcp"]
+	if !ok || len(m) == 0 {
+		t.Fatalf("failed to get HTTP host port mapping from openfga container")
+	}
+	httpPort := m[0].HostPort
+
+	m, ok = ports["8081/tcp"]
+	if !ok || len(m) == 0 {
+		t.Fatalf("failed to get grpc host port mapping from openfga container")
+	}
+	grpcPort := m[0].HostPort
+
+	timeoutCtx, cancel := context.WithTimeout(ctx, 3*time.Second)
+	defer cancel()
+
+	creds := insecure.NewCredentials()
+
+	dialOpts := []grpc.DialOption{
+		grpc.WithBlock(),
+		grpc.WithTransportCredentials(creds),
+		grpc.WithConnectParams(grpc.ConnectParams{Backoff: grpcbackoff.DefaultConfig}),
+	}
+
+	conn, err := grpc.DialContext(
+		timeoutCtx,
+		fmt.Sprintf("localhost:%s", grpcPort),
+		dialOpts...,
+	)
+	require.NoError(t, err)
+	defer conn.Close()
+
+	client := healthv1pb.NewHealthClient(conn)
+
+	backoffPolicy := backoff.NewExponentialBackOff()
+	backoffPolicy.MaxElapsedTime = 30 * time.Second
+
+	err = backoff.Retry(func() error {
+		timeoutCtx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
+		defer cancel()
+
+		resp, err := client.Check(timeoutCtx, &healthv1pb.HealthCheckRequest{
+			Service: openfgapb.OpenFGAService_ServiceDesc.ServiceName,
+		})
+		if err != nil {
+			return err
+		}
+
+		if resp.GetStatus() != healthv1pb.HealthCheckResponse_SERVING {
+			return fmt.Errorf("not serving")
+		}
+
+		return nil
+	},
+		backoffPolicy,
+	)
+	require.NoError(t, err)
+
+	return &serverHandle{
+		grpcPort: grpcPort,
+		httpPort: httpPort,
+		cleanup:  stopContainer,
+	}, nil
+}
+
+func TestFunctionalGRPC(t *testing.T) {
+
+	require := require.New(t)
+
+	// tester can be shared across tests that aren't impacted
+	// by shared state
+	tester, err := newOpenFGATester(t)
+	require.NoError(err)
+	defer tester.Cleanup()
+
+	t.Run("TestCreateStore", func(t *testing.T) { GRPCCreateStoreTest(t, tester) })
```

Comment:
i'm confused as to why we have `GRPCCreateStoreTest` and `HTTPCreateStoreTest` - are we going to duplicate the code? 🤯

Reply:
They are both part of the API surface, and given that we use middleware that differs between the two I think it's important we test the functional API behavior of both :).

So yeah, we'll test both.